### PR TITLE
[SPARK-6844][SQL] Clean up accumulators used in InMemoryRelation when it is uncached

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-6844

### CODE REVIEW

**Comment by marmbrus** [comment]
Thanks for working on this and for adding a test!  I have two issues with the way this is being implemented:
 - We are still leaking accumulators on a per table scan basis, and so you are still going to have a problem if you never explicitly uncache the table
 - It seems like we should be able to do this without adding a global hash map.  In particular, when you are uncaching you have a handle to the InMemoryRelation.  Perhaps that class should define an `uncache` method on this class that takes care of both `unpersist`ing the data and unregistering the accumulators.

The ones inside of the table scan are clearly harder to deal with, but perhaps we can just make those lazy vals that are only initialized during testing due to some conf being set?

**Comment by viirya** [comment]
Thanks for commenting.

For the first issue, I have registered the accumulators of table scan `InMemoryColumnarTableScan`.  If users explicitly uncache the table, these accumulators can be released as I tested. If the table is not explicitly uncached, these accumulators will be leaked as you said. Using a configuration to initialize them sounds good. I will try it.

For the second one, I have tried doing this without a global hash map. However, because the cache manager will call `InMemoryRelation.withOutput` to create a new `InMemoryRelation` each time replacing the logical plan with cached version. The accumulators will be created in these `InMemoryRelation`s. But the cache manager only keeps the original `InMemoryRelation` and we uncache based on it. So I add this global hash map to deal with this problem.

**Comment by marmbrus** [comment]
I see, thanks for clarifying.  It seems bad that we are creating accumulators per copy anyway then.  Can we just pass in the accumulator with the buffers and statistics instead so that there is only ever one copy?

Honestly, the best solution would probably be to make the `InMemoryRelation` a `BaseRelation` so that it does not have to handle `newInstance` calls at all.  Unfortunately this API did not exist when that class was created.  However, I'm not sure its worth refactoring at this point.  This code was ported from Shark and I'd like to get rid of it completely in the next release or two.

For the second part, it would be great if we could just get rid of the accumulators completely. It feels bad to have all this extra logic only for a test.  Can you think of any other way that we could test this?

**Comment by viirya** [comment]
i just found that we can only allocate the accumulator for first one `InMemoryRelation` instead of for all following `InMemoryRelation`. Because the following relations only need to use the pre-computed statistics of the first relation.

The accumulators in `InMemoryColumnarTableScan` are harder to completely removed if we want to collect these statistics, because these statistics are computed during the RDD operations. Accumulator seems the only method we can pass the information out. Currently I still apply a configuration to control if we want to allocate accumulators. But now I put these accumulators in an array in `InMemoryRelation`, instead of a global map before. Of course this array needs to be propagated to following relations because `InMemoryColumnarTableScan` is depending on later relations not the original one.

**Comment by marmbrus** [comment]
For `InMemoryColumnarTableScan` if they are only every allocated during testing, I think it is okay if they leak.

**Comment by viirya** [comment]
ok. updated.

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

File: sql/core/src/test/scala/org/apache/spark/sql/columnar/PartitionBatchPruningSuite.scala [hunk 1]
```diff
@@ -39,6 +39,8 @@ class PartitionBatchPruningSuite extends FunSuite with BeforeAndAfterAll with Be
 
     // Enable in-memory partition pruning
     setConf(SQLConf.IN_MEMORY_PARTITION_PRUNING, "true")
+    // Enable in-memory table scan accumulators
+    setConf("spark.sql.inMemoryTableScanStatistics.enable", "true")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/columnar/PartitionBatchPruningSuite.scala]: Why not setting `spark.sql.inMemoryTableScanStatistics.enable` back `false`?

File: sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala [hunk 2]
```diff
@@ -297,4 +298,21 @@ class CachedTableSuite extends QueryTest {
     sql("Clear CACHE")
     assert(cacheManager.isEmpty)
   }
+
+  test("Clear accumulators when uncacheTable to prevent memory leaking") {
+    val accsSize = Accumulators.originals.size
+
+    sql("SELECT key FROM testData LIMIT 10").registerTempTable("t1")
+    sql("SELECT key FROM testData LIMIT 5").registerTempTable("t2")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala]: We also need to drop these two temp views after tests

**Pull request closed without merging**